File: py/day20.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

import re

lut = {"#": 1, ".": 0}

def parse(path):
    algo, image = [], []
    with open(path) as f:
        for line in f.readlines():
            if len(line.strip()) == 512:
                algo = [lut[c] for c in re.findall(r"\S", line)]
            elif line.strip():
                image.append([lut[c] for c in line.strip()])
    return {'algo': algo, 'image': image}

sample_input = parse('../resources/day20_ex.txt')
puzzle_input = parse('../resources/day20.txt')

def square(y, x):
    return [(y-1,x-1), (y-1,x), (y-1,x+1),
            (y,x-1),    (y,x),    (y,x+1),
            (y+1,x-1), (y+1,x), (y+1,x+1)]

def get_index(image, y, x, dummy=0):
    height, width = len(image), len(image[0])
    bits = [image[y][x] if 0 <= y < height and 0 <= x < width else dummy
            for (y, x) in square(y, x)]
    return int("".join([str(b) for b in bits]), 2)

def solve(iterations, algo, image):
    for i in range(iterations):
        dummy = 0 if algo[0] == 0 else i % 2
        height, width = len(image), len(image[0])
        image = [[algo[get_index(image, y, x, dummy)] for x in range(-1, width+1)]
                                                      for y in range(-1, height+1)]
    return sum([sum([1 if x == 1 else 0 for x in row]) for row in image])

print("Day 15\n"
      "[Part 1] Example: %d, Puzzle: %d\n"
      "[Part 2] Example: %d, Puzzle: %d"
      %(solve(2, **sample_input), solve(2, **puzzle_input),
        solve(50, **sample_input), solve(50, **puzzle_input)))

File: src/aoc2021/day20.clj
(ns aoc2021.day20
  (:require [clojure.java.io :as io]))

(comment
 "Day 20: Trench Map")

(defn parse
  [data]
  (let [lut {\# 1 \. 0}]
    {:algo           (mapv lut (first data))
     :times-enhanced 0
     :image          (->> (drop 2 data)
                          (mapv (partial mapv lut)))}))

(def sample-input
  (-> "../resources/day20_ex.txt" io/resource io/reader line-seq parse))

(def puzzle-input
  (-> "../resources/day20.txt" io/resource io/reader line-seq parse))

(defn square
  [[y x]]
  [[(dec y) (dec x)] [(dec y) x] [(dec y) (inc x)]
   [y (dec x)]       [y x]       [y (inc x)]
   [(inc y) (dec x)] [(inc y) x] [(inc y) (inc x)]])

(defn bin->dec
  [x]
  (Integer/parseInt x 2))

(defn index
  [image [y x] dummy]
  (->> (square [y x])
       (map #(get-in image % dummy))
       (apply str)
       bin->dec))

(defn enhance
  [{:keys [algo image times-enhanced] :as state}]
  ;; Tricky! If the first bit in the algo is 1, then all of the "empty" spaces
  ;; in the infinite region outside the image become 1 upon every other enhancement.
  (let [dummy (if (zero? (first algo)) 0
                (mod times-enhanced 2))]
    (-> state
        (update :times-enhanced inc)
        (assoc :image
          (let [height (count image)
                width  (count (first image))]
            (vec (for [y (range -1 (inc height))]
                   (vec (for [x (range -1 (inc width))]
                          (nth algo (index image [y x] dummy)))))))))))

(defn solve
  [{:keys [times-enhanced image] :as data} iterations]
  (if (= times-enhanced iterations)
    (-> image flatten frequencies (get 1))
    (recur (enhance data) iterations)))

(time
 (do
  (println "Day 20: Trench Map")
  (println "[Part 1] Sample:" (solve sample-input 2) "Puzzle:" (solve puzzle-input 2))
  (println "[Part 2] Sample:" (solve sample-input 50) "Puzzle:" (solve puzzle-input 50))))
